### PR TITLE
Switched default document type to _doc

### PR DESCRIPTION
**Pull request by mtorromeo**
I would like to propose a breaking change that would allow the library to work with Elasticsearch 7 where the document type was removed after being deprecated in version 6.

The change consists mostly of defaulting to the `_doc` type instead of `doc`. By doing this, most ES routes work as intended.

I'm using this in one of my applications that uses the generated mapping and indexes documents just fine.

Note that while it would be possible to manually specify the `_doc` type everywhere, the mapping API should still be changed from `PUT /my-index/_mappings/_doc` to `PUT /my-index/_mapping`.

### CODE REVIEW

**Comment by mtorromeo** [comment]
Related issue: #337

**Comment by KodrAus** [comment]
Thanks for doing this @mtorromeo! The build failure is coming from our sample app: https://travis-ci.org/elastic-rs/elastic/jobs/532705620#L1764-L1767. We just need to update the test here: https://github.com/elastic-rs/elastic/blob/master/examples/account_sample/src/ops/commands/put_bulk_accounts.rs#L87

**Comment by mtorromeo** [comment]
@KodrAus I updated the `BulkUrlParams::url` function with special-case handling of the `_doc` type.

There are probably other APIs that should be fixed like this because not all of them work with the `/_doc/` part in the URL but I have not tested many.

**Comment by KodrAus** [comment]
Ah, those urls are all automatically generated [by this project](https://github.com/elastic-rs/elastic/tree/master/src/requests_codegen) so we can’t just change those files manually.

It looks like the [REST api spec](https://github.com/elastic/elasticsearch/tree/master/rest-api-spec/src/main/resources/rest-api-spec/api) these are generated from still have the `{type}` parameter in routes.

If we want to avoid setting the route if the type is `_doc` then I think the best place for that logic is on the client request builders. As an example, [here’s where we convert a bulk request into the lower-level generated one](https://github.com/elastic-rs/elastic/blob/master/src/elastic/src/client/requests/bulk/mod.rs#L528) and could decide how to treat the type there. All of the client requests have an `into_request` method we can use.

Does that make sense?

**Comment by mtorromeo** [comment]
Ok, I get it now, thanks for the clear explanation and the links.

We should be fine keeping all the `/_doc/` parts in the URLs for now but those are bound to be removed sooner or later and it doesn't hurt to start fixing them if possible. This would also avoid filling the deprecation log.

I'll fix the PR probably in the next few days.

**Comment by mtorromeo** [comment]
@KodrAur I'm having issues working with `requests_codegen`.

To update `genned.rs` for ES7 I think I should update the specs in `src/requests_codegen/spec` and the run `cargo run > ../requests/src/genned.rs` in `requests_codegen`?

The problem is that, even without touching the specs folder, when I run the codegen, the generated code is wrongly formatted. https://gist.github.com/mtorromeo/5b06a68d58fcd25a8a00daafd2639761

The code is mostly on one line with extra spaces everywhere.

I need to update it to refactor the PR since some functions are not present, like `IndicesPutMappingRequest::for_index`.

**Comment by mtorromeo** [comment]
I updated the PR with the file `genned.rs` still modified by hand but only adding the missing `IndicesPutMappingRequest::for_index` function that should be the same as the auto-generated version if I could make `requests_codegen` work as intended.

File: src/elastic/src/client/requests/bulk/mod.rs [hunk 1]
```diff
@@ -524,8 +524,11 @@ where
         let body = self.body.try_into_inner()?;
 
         match (self.index, self.ty) {
-            (Some(index), None) => Ok(BulkRequest::for_index(index, body)),
-            (Some(index), Some(ty)) => Ok(BulkRequest::for_index_ty(index, ty, body)),
+            (Some(index), ty) => match ty {
+                None => Ok(BulkRequest::for_index(index, body)),
+                Some(ref ty) if &ty[..] == "_doc" => Ok(BulkRequest::for_index(index, body)),
```
Comment (KodrAus) [inline src/elastic/src/client/requests/bulk/mod.rs]: We could actually use the `DEFAULT_DOC_TYPE` constant here. It's available under `elastic::types::document::DEFAULT_DOC_TYPE`.

I think after this I'll take the plunge proper and remove the multiple crates approach. Then we could just add an inherent method to `Type` that checks whether it's the default or not.

**Comment by KodrAus** [comment]
> To update genned.rs for ES7 I think I should update the specs in src/requests_codegen/spec and then run cargo run > ../requests/src/genned.rs in requests_codegen?

That's right. We pull that spec in from upstream though, so we can't really make manual modifications to it. There is a place in `requests_codegen` where we can apply tweaks though. I'm happy to take your manual edits and work them into the generation process, since it is a bit nasty.

> The problem is that, even without touching the specs folder, when I run the codegen, the generated code is wrongly formatted.

Yes it just dumps out unformatted Rust code, then we need to come back and run `cargo fmt` over.

**Comment by mtorromeo** [comment]
Yes, I understood that the spec files are from upstream but they are not up-to-date with version 7 of Elasticsearch. 

If it's ok, I'll pull the latest spec files from upstream and make it part of this PR.

**Comment by mtorromeo** [comment]
It worked almost perfectly right away but I had to fix the generated code: https://github.com/elastic-rs/elastic/pull/356/commits/0e4266c9600f82142375e9f3b3c4235f513ec33a

It seems the upstream specs are missing the leading slash on some of the paths.

This is broken on both the v7.1.1 tag that I used: https://github.com/elastic/elasticsearch/blob/7a013ded4033c6cc60ad2078f3e1f18666bec47a/rest-api-spec/src/main/resources/rest-api-spec/api/indices.put_mapping.json#L7

and on master: https://github.com/elastic/elasticsearch/blob/b3a778093c056fe320c800ab7d60f094077f0103/rest-api-spec/src/main/resources/rest-api-spec/api/indices.put_mapping.json#L36

**Comment by mtorromeo** [comment]
It's also unfortunate that the generated code has the code blocks in a very different order so genned.rs is basically a complete rewrite in git.

**Comment by KodrAus** [comment]
Thanks for your patience on this one @mtorromeo!

This looks good to me so I'll merge it in and get everything ready for another pre-release.

**Pull request merged**